### ISSUE TITLE
useless className warning

### ISSUE THREAD

**Issue by B-Reif** [styled-components/styled-components#2159]
I'm using styled-components to style a `rc-tooltip`, which takes the property `overlayClassName` instead of `className` in order to style the tooltip's overlay. My code contains a very basic wrapper to map the prop:

```
const TooltipStyleWrapper = props => {
	const { className, ...rest } = props;
	return <Tooltip overlayClassName={className} {...rest} />;
};
```

When this component mounts, styled-components issues the className warning:
> It looks like you've wrapped styled() around your React component (TooltipStyleWrapper), but the className prop is not being passed down to a child. No styles will be rendered unless className is composed within your React component.

In truth, I _am_ passing down `className`, but by a different name. I did a brief search for this issue and saw PR #2156 for adding another prop to suppress this warning.

I don't think it's appropriate for styled-components to pry into my components to see how I'm using the provided `className`, and I _really_ don't think it's appropriate that I should have to add another obtuse prop to suppress this warning. The above code already exists as a bridge between `sc` and `rc-tooltip`, and I very much don't want to go to even further lengths to build this bridge. If a warning is going to exist, it should be smart enough only to warn me when something is _actually wrong_.

**Comment by Fer0x** [comment]
This warning is not about that you are using className prop name in your code. It search in nesting children to sure that you have a real DOM node which uses passed classnames.

**Comment by Fer0x** [comment]
There is only one problem with this warning when asynchronous rendering is used. I have some ideas and will try to make some solution to handle it.

**Comment by B-Reif** [comment]
The className here is ultimately rendered in a node that's set up with a portal, so it's both asynchronous (drawn later, when someone mouses over the tooltip) and detached (rendered elsewhere than the DOM tree).

If it's only searching the nested DOM nodes, it won't find content rendered in a portal.

**Comment by probablyup** [comment]
You'll be able to opt out of this warning for edge cases like you've described: https://github.com/styled-components/styled-components/pull/2156

Probably will cut a release tomorrow.

**Comment by B-Reif** [comment]
I'm really not happy with the need to add another prop just to suppress this warning, and I don't really view deferred or relocated rendering as an 'edge case'. I should be able to render the className later, or somewhere else, or not at all. Once the className is handed off,  it's not styled-component's business how, if, or when I choose to use it.

**Comment by probablyup** [comment]
> Once the className is handed off, it's not styled-components's business how, if, or when I choose to use it.

Quite true, however we get enough issues opened on the repo for this particular problem that require our time to triage and respond to that adding the warning is a better solution for our productivity than doing nothing. The suppression prop is an escape hatch if you know what you're doing, and you clearly do.

**Comment by B-Reif** [comment]
That's fair! I think this kind of speaks to a bigger problem, which is that `styled(ReactComponent)` is a leaky abstraction that still requires developers to understand a `className` is being created and applied. This warning is an ok band-aid for the leak, but I wonder if there's a better way to present what sc is doing.

**Comment by probablyup** [comment]
With the upcoming "hooks" pattern that might be possible. There have already been some experiments internally on the subject. Keep an eye on the repo in the next few months, there will probably will be more to say on the subject as the RFC evolves.

**Comment by B-Reif** [comment]
I was just about to comment that a hook API would be more appropriate and you beat me to it!

**Comment by B-Reif** [comment]
One last thought: it might be nice to set a `didWarn` flag or something so that you only warn me once. I mount a fair number of tooltips and got the warning many times on one render.

**Comment by probablyup** [comment]
cc @Fer0x - can you add a guard so that each component only emits the warning once? We can do a simple reference equality test using an ES6 Map and set a flag once the warning has gone out once.

**Comment by Fer0x** [comment]
@probablyup sure, will add this today
